### PR TITLE
drivers: sensor: ams-tsl2520: add initial driver files

### PR DESCRIPTION
**Pull request by yunheQin**
This commit adds ams,tsl2520.yaml to dts/bindings/sensor

This commit adds initial driver files of ams-tsl2520 to
support the highly sensitive ambient light sensor

This commit modifies the CMakeLists.txt and Kconfig files
in the folder "drivers/sensor/" to support ams-tsl2520

This commit modifies the i2c.dtsi file in the tests/drivers/
build_all/sensor folder

### CODE REVIEW

File: drivers/sensor/tsl2520/tsl2520.c [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+	tmp = cfg->sample_time;
+	ret = i2c_reg_write_byte_dt(&cfg->i2c_spec, TSL2520_SAMPLE_TIME0, tmp);
+	if (ret < 0) {
+		LOG_ERR("Failed setting sample time");
+		return ret;
+	}
+
+	temp = cfg->als_nr_samples;
+	ret = i2c_reg_write_byte_dt(&cfg->i2c_spec, TSL2520_ALS_NR_SAMPLES0, LO_BYTE (temp));
```
Comment (marcusfolkesson) [inline drivers/sensor/tsl2520/tsl2520.c:123]: Skip LO_BYTE/HI_BYTE and use i2c_burst_write_dt() instead.
The device supports burst writes and even recommends it.

"All 16-bit fields have a latching scheme for reading and writing.
In general it is recommended to use I²C bursts whenever
possible, "

- https://ams.com/documents/20143/4410897/TSL2520_DS001068_3-00.pdf

File: drivers/sensor/tsl2520/tsl2520.c [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+		LOG_ERR("Failed setting als nr sample time");
+		return ret;
+	}
+
+	temp = cfg->agc_nr_samples;
+	ret = i2c_reg_write_byte_dt(&cfg->i2c_spec, TSL2520_AGC_NR_SAMPLES_LOW, LO_BYTE (temp));
+	ret = i2c_reg_write_byte_dt(&cfg->i2c_spec, TSL2520_AGC_NR_SAMPLES_HIGH, HI_BYTE(temp));
+	if (ret < 0) {
+		LOG_ERR("Failed setting agc nr sample time");
+		return ret;
```
Comment (marcusfolkesson) [inline drivers/sensor/tsl2520/tsl2520.c:135]: Remove this return...

File: drivers/sensor/tsl2520/tsl2520.c [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+	}
+
+	return 0;
```
Comment (marcusfolkesson) [inline drivers/sensor/tsl2520/tsl2520.c:138]: ...and change this to 
return ret;

File: drivers/sensor/tsl2520/tsl2520.c [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+		/* In order to update ALS Data Registers ALS_STATUS must be read first */
+		(void)i2c_reg_read_byte_dt(&cfg->i2c_spec, TSL2520_REG_STATUS, &reg_status);
+
+		ret = i2c_reg_read_byte_dt(&cfg->i2c_spec, TSL2520_REG_STATUS, &reg_status);
+		if (ret < 0) {
+			LOG_ERR("Failed reading chip status");
+			return -EIO;
+		}
+		ret = i2c_reg_read_byte_dt(&cfg->i2c_spec, TSL2520_REG_ALS_DATA0_LOW, &read_low_byte);
```
Comment (marcusfolkesson) [inline drivers/sensor/tsl2520/tsl2520.c:54]: i2c_burst_read_dt()

File: drivers/sensor/tsl2520/tsl2520.c [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+		}
+		ret = i2c_reg_read_byte_dt(&cfg->i2c_spec, TSL2520_REG_ALS_DATA0_HIGH, &read_high_byte);
+		if (ret < 0) {
+			LOG_ERR("Could not fetch als sensor value");
+			return -EIO;
+		}
+		data->als_data0 = (read_high_byte << 8) | read_low_byte;
+		LOG_DBG("data0 value: %X ",data->als_data0 );
+
+		ret = i2c_reg_read_byte_dt(&cfg->i2c_spec, TSL2520_REG_ALS_DATA1_LOW, &read_low_byte);
```
Comment (marcusfolkesson) [inline drivers/sensor/tsl2520/tsl2520.c:67]: i2c_burst_read_dt()

File: drivers/sensor/tsl2520/tsl2520.c [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+	}
+	
+	return ret;
+}
+
+static int tsl2520_channel_get(const struct device *dev, enum sensor_channel chan,
+			       struct sensor_value *val)
+{
+	struct tsl2520_data *data = dev->data;
+	int ret = 0;
```
Comment (marcusfolkesson) [inline drivers/sensor/tsl2520/tsl2520.c:93]: Remove ret

File: drivers/sensor/tsl2520/tsl2520.c [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+	if (chan == SENSOR_CHAN_ALL || chan == SENSOR_CHAN_LIGHT || chan == SENSOR_CHAN_IR) {
+		val->val1 = data->als_data0;
+		val->val2 = data->als_data1;
+		return ret;
```
Comment (marcusfolkesson) [inline drivers/sensor/tsl2520/tsl2520.c:97]: return 0;

File: drivers/sensor/tsl2520/tsl2520.c [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+static int tsl2520_pm_action(const struct device *dev, enum pm_device_action action)
+{
+
+	const struct tsl2520_config *cfg = dev->config;
+	struct tsl2520_data *data = dev->data;
+	int ret = 0;
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		ret = i2c_reg_update_byte_dt(&cfg->i2c_spec, TSL2520_ENABLE_ADDR,
```
Comment (marcusfolkesson) [inline drivers/sensor/tsl2520/tsl2520.c:190]: return i2c_reg_update_byte_dt(...

File: drivers/sensor/tsl2520/tsl2520.c [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+	.channel_get = tsl2520_channel_get,
+};
+
+#ifdef CONFIG_PM_DEVICE
+static int tsl2520_pm_action(const struct device *dev, enum pm_device_action action)
+{
+
+	const struct tsl2520_config *cfg = dev->config;
+	struct tsl2520_data *data = dev->data;
+	int ret = 0;
```
Comment (marcusfolkesson) [inline drivers/sensor/tsl2520/tsl2520.c:186]: remove ret

File: drivers/sensor/tsl2520/tsl2520.c [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+
+	switch (action) {
+	case PM_DEVICE_ACTION_RESUME:
+		ret = i2c_reg_update_byte_dt(&cfg->i2c_spec, TSL2520_ENABLE_ADDR,
+					    TSL2520_ENABLE_MASK, TSL2520_ENABLE_MASK);
+		break;
+	case PM_DEVICE_ACTION_SUSPEND:
+		ret = i2c_reg_update_byte_dt(&cfg->i2c_spec, TSL2520_ENABLE_ADDR,
```
Comment (marcusfolkesson) [inline drivers/sensor/tsl2520/tsl2520.c:194]: return i2c_reg_upd...

File: drivers/sensor/tsl2520/tsl2520.c [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+					    TSL2520_ENABLE_MASK, TSL2520_ENABLE_DISABLE);
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	return ret;
```
Comment (marcusfolkesson) [inline drivers/sensor/tsl2520/tsl2520.c:201]: return 0;

File: drivers/sensor/tsl2520/tsl2520.h [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+#define TSL2520_MEAS_SEQR_STEP3_MOD_PHDX_SMUX_L		0xE2
+#define TSL2520_MEAS_SEQR_STEP3_MOD_PHDX_SMUX_H		0xE3
+
+#define TSL2520_MOD_CALIB_CFG0		0xE4
+#define TSL2520_MOD_CALIB_CFG2		0xE6
+
+
+/* ENABLE(0x80: 0x00): Reserved:7:2 | AEN:1 | PON:0 */
+#define TSL2520_ENABLE_ADDR 0x80
+#define TSL2520_ENABLE_MASK (BIT(1) | BIT(0))
```
Comment (marcusfolkesson) [inline drivers/sensor/tsl2520/tsl2520.h:96]: #define TSL2520_ENABLE_MASK GENMASK(1,0)

File: drivers/sensor/tsl2520/tsl2520.h [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+#define TSL2520_ENABLE_AEN_PON (BIT(1) | BIT(0))
```
Comment (marcusfolkesson) [inline drivers/sensor/tsl2520/tsl2520.h:97]: GENMASK()

File: drivers/sensor/tsl2520/tsl2520.c [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+#include <zephyr/sys/__assert.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/sys/util.h>
+
+#define TSL2520_CHIP_ID 0x5C
+
+#define LO_BYTE(X)	(X & 0xFF)
```
Comment (marcusfolkesson) [inline drivers/sensor/tsl2520/tsl2520.c:27]: Remove these

File: drivers/sensor/tsl2520/tsl2520.c [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+	temp = cfg->als_nr_samples;
+	ret = i2c_reg_write_byte_dt(&cfg->i2c_spec, TSL2520_ALS_NR_SAMPLES0, LO_BYTE (temp));
+	ret = i2c_reg_write_byte_dt(&cfg->i2c_spec, TSL2520_ALS_NR_SAMPLES1, HI_BYTE(temp));
+	if (ret < 0) {
+		LOG_ERR("Failed setting als nr sample time");
+		return ret;
+	}
+
+	temp = cfg->agc_nr_samples;
+	ret = i2c_reg_write_byte_dt(&cfg->i2c_spec, TSL2520_AGC_NR_SAMPLES_LOW, LO_BYTE (temp));
```
Comment (marcusfolkesson) [inline drivers/sensor/tsl2520/tsl2520.c:131]: i2c_burst_write_dt

**Review by marcusfolkesson — CHANGES_REQUESTED** [review]
Hi!

Thank you for this PR!
It looks good overall.
Just a few nitpicking comments as I'm grumpy on a Monday morning :-)

File: drivers/sensor/tsl2520/tsl2520.c (hunk 4 above)
Comment (yunheQin) [reply drivers/sensor/tsl2520/tsl2520.c:54]: hello @marcusfolkesson: Thank you very much for your review, but here I have some different thoughts, there are 2 independent regesisters: TSL2520_REG_ALS_DATA0_LOW and TSL2520_REG_ALS_DATA0_HIGH,  the address of them are different ,but the function of the "i2c_burst_read_dt()" is to _Read multiple bytes from an internal address of an I2C device. This is equivalent to: i2c_burst_read(spec->bus, spec->addr, start_addr, buf, num_bytes);_ , at the same time I also did some test to print the read values from these two registers the value are different, using the function "i2c_burst_read_dt()" and "i2c_reg_read_byte_dt".

Comment (yunheQin) [reply drivers/sensor/tsl2520/tsl2520.c:54]: to help understanding, I will paste the print information which read value send from UART.
image
the read format is LOG_DBG("burst_read%X,%X,%X",le16_buffer,read_low_byte,read_high_byte);

File: drivers/sensor/tsl2520/tsl2520.c [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+		if (ret < 0) {
+			LOG_ERR("Failed reading chip status");
+			return -EIO;
+		}
+		ret = i2c_burst_read_dt(&cfg->i2c_spec,TSL2520_REG_ALS_DATA0_LOW,(uint8_t *)&le16_buffer, sizeof(le16_buffer));
+		if (ret < 0) {
+			LOG_ERR("Could not fetch als sensor value");
+			return -EIO;
+		}
+		LOG_DBG("data0 value: %X ",data->als_data0 );
```
Comment (kartben) [inline drivers/sensor/tsl2520/tsl2520.c:55]:
```suggestion
```

File: drivers/sensor/tsl2520/tsl2520.c [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+
+		ret = i2c_burst_read_dt(&cfg->i2c_spec,TSL2520_REG_ALS_DATA1_LOW,(uint8_t *)&le16_buffer, sizeof(le16_buffer));
+		if (ret < 0) {
+			LOG_ERR("Could not fetch als sensor value");
+			return -EIO;
+		}
+		data->als_data0 = sys_le16_to_cpu(le16_buffer);
+		LOG_DBG("data1 value: %X ",data->als_data1 );
```
Comment (kartben) [inline drivers/sensor/tsl2520/tsl2520.c:63]:
```suggestion
		data->als_data1 = sys_le16_to_cpu(le16_buffer);
```

File: drivers/sensor/tsl2520/tsl2520.c [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+
+#include "tsl2520.h"
+
+#include <stdlib.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/pm/device.h>
```
Comment (kartben) [inline drivers/sensor/tsl2520/tsl2520.c:17]:
```suggestion
```

File: dts/bindings/sensor/ams,tsl2520.yaml [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+      and is calculated as: agc_atime = (AGC_NR_SAMPLES+1) x
+      (SAMPLE_TIME+1) x 1.388889μs
+
+  als-gains:
+    type: int
+    default: 128
+    description: |
+      Defines the gain of modulator for the measurement
+      sequencer steps. 
+      The gain is also updated by the AGC, if activated
```
Comment (kartben) [inline dts/bindings/sensor/ams,tsl2520.yaml:54]: Default values need to be documented
https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#dt-bindings-default-rules

**Review by kartben — CHANGES_REQUESTED** [review]
Few comments

File: drivers/sensor/tsl2520/tsl2520.c [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+	(void)i2c_reg_read_byte_dt(&cfg->i2c_spec, TSL2520_REG_ID, &chip_id);
+
+	ret = i2c_reg_read_byte_dt(&cfg->i2c_spec, TSL2520_REG_ID, &chip_id);
+	if (ret < 0) {
+		LOG_ERR("Failed reading chip id");
+		return ret;
+	}
+
+	if (chip_id != TSL2520_CHIP_ID) {
+		LOG_ERR("Chip id is invalid! Device @%02x is no TMD2620!", cfg->i2c_spec.addr);
```
Comment (RichardSWheatley) [inline drivers/sensor/tsl2520/tsl2520.c:144]: Wrong chip mentioned in the LOG_ERR

File: drivers/sensor/tsl2520/tsl2520.c [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+	tmp = cfg->sample_time;
+	ret = i2c_reg_write_byte_dt(&cfg->i2c_spec, TSL2520_SAMPLE_TIME0, tmp);
+	if (ret < 0) {
+		LOG_ERR("Failed setting sample time");
+		return ret;
+	}
+
+	temp = cfg->als_nr_samples;
+	ret = i2c_burst_write_dt(&cfg->i2c_spec, TSL2520_ALS_NR_SAMPLES0, (uint8_t *)&temp,
+				 sizeof(temp));
```
Comment (aasinclair) [inline drivers/sensor/tsl2520/tsl2520.c:104]: i2c_burst_write does not work correctly, so should be avoided (https://github.com/zephyrproject-rtos/zephyr/issues/20154).
Also you need to explicitly set the endianess of the data in the message, e.g:

```suggestion
	uint8_t buf[3] = {TSL2520_ALS_NR_SAMPLES0};
	sys_put_le16(cfg->als_nr_samples, &buf[1]);
	int ret = i2c_write_dt(&cfg->i2c_spec, buf, sizeof(buf));
```

File: drivers/sensor/tsl2520/tsl2520.c [hunk 22]
```diff
@@ -0,0 +1,198 @@
...
+	uint8_t chip_id;
+	int ret;
+
+	if (!i2c_is_ready_dt(&cfg->i2c_spec)) {
+		LOG_ERR("I2C dev %s not ready", cfg->i2c_spec.bus->name);
+		return -ENODEV;
+	}
+	/* trying to read the id twice, as the sensor does not answer the first request */
+	/* because of this no return code is checked in this line */
+	(void)i2c_reg_read_byte_dt(&cfg->i2c_spec, TSL2520_REG_ID, &chip_id);
```
Comment (aasinclair) [inline drivers/sensor/tsl2520/tsl2520.c:132]: Is this expected behaviour for this sensor?

File: drivers/sensor/tsl2520/tsl2520.c [hunk 23]
```diff
@@ -0,0 +1,198 @@
...
+
+#define TSL2520_DEFINE(inst)                                                                       \
+	static struct tsl2520_data tsl2520_prv_data_##inst;                                        \
+	static const struct tsl2520_config tsl2520_config_##inst = {                               \
+		.i2c_spec = I2C_DT_SPEC_INST_GET(inst),                                            \
+		.int_gpio = GPIO_DT_SPEC_INST_GET(inst, int_gpios),                                \
+		.sample_time = DT_INST_PROP(inst, sample_time),                                    \
+		.als_nr_samples = DT_INST_PROP(inst, als_nr_samples),                              \
+		.agc_nr_samples = DT_INST_PROP(inst, agc_nr_samples),                              \
+		.als_gains = DT_INST_PROP(inst, als_gains),                                        \
```
Comment (aasinclair) [inline drivers/sensor/tsl2520/tsl2520.c:190]: Is this used anywhere?

**Review by aasinclair — CHANGES_REQUESTED** [review]

**Pull request closed without merging**